registry/client/layer.go

```diff
@@ -0,0 +1,178 @@
+package client
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/digest"
+)
+
+type httpLayer struct {
+	*layers
+
+	size      int64
+	digest    digest.Digest
+	createdAt time.Time
+
+	rc     io.ReadCloser // remote read closer
+	brd    *bufio.Reader // internal buffered io
+	offset int64
+	err    error
+}
+
+func (hl *httpLayer) CreatedAt() time.Time {
+	return hl.createdAt
+}
+
+func (hl *httpLayer) Digest() digest.Digest {
+	return hl.digest
+}
+
+func (hl *httpLayer) Read(p []byte) (n int, err error) {
+	if hl.err != nil {
+		return 0, hl.err
+	}
+
+	rd, err := hl.reader()
+	if err != nil {
+		return 0, err
+	}
+
+	n, err = rd.Read(p)
+	hl.offset += int64(n)
+
+	// Simulate io.EOF error if we reach filesize.
+	if err == nil && hl.offset >= hl.size {
+		err = io.EOF
+	}
+
+	return n, err
+}
+
+func (hl *httpLayer) Seek(offset int64, whence int) (int64, error) {
+	if hl.err != nil {
+		return 0, hl.err
+	}
+
+	var err error
+	newOffset := hl.offset
+
+	switch whence {
+	case os.SEEK_CUR:
+		newOffset += int64(offset)
+	case os.SEEK_END:
+		newOffset = hl.size + int64(offset)
+	case os.SEEK_SET:
+		newOffset = int64(offset)
+	}
+
+	if newOffset < 0 {
+		err = fmt.Errorf("cannot seek to negative position")
+	} else {
+		if hl.offset != newOffset {
+			hl.reset()
+		}
+
+		// No problems, set the offset.
+		hl.offset = newOffset
+	}
+
+	return hl.offset, err
+}
+
+func (hl *httpLayer) Close() error {
+	if hl.err != nil {
+		return hl.err
+	}
+
+	// close and release reader chain
+	if hl.rc != nil {
+		hl.rc.Close()
+	}
+
+	hl.rc = nil
+	hl.brd = nil
+
+	hl.err = fmt.Errorf("httpLayer: closed")
+
+	return nil
+}
+
+func (hl *httpLayer) reset() {
+	if hl.err != nil {
+		return
+	}
+	if hl.rc != nil {
+		hl.rc.Close()
+		hl.rc = nil
+	}
+}
+
+func (hl *httpLayer) reader() (io.Reader, error) {
+	if hl.err != nil {
+		return nil, hl.err
+	}
+
+	if hl.rc != nil {
+		return hl.brd, nil
+	}
+
+	// If the offset is great than or equal to size, return a empty, noop reader.
+	if hl.offset >= hl.size {
+		return ioutil.NopCloser(bytes.NewReader([]byte{})), nil
+	}
+
+	blobURL, err := hl.ub.BuildBlobURL(hl.name, hl.digest)
```

Comment:
@dmcgowan @stevvooe 

It would be so much better to have an exported constructor that takes a static url string and an http client to make it very generic so that I can reuse it:

```
func NewHttpReadSeeker(client *http.Client, url string) io.ReadSeeker {
    return *httpLayer{...}
}
```

you shouldn't need to embed `*layers`